crates/common/mqtt_channel/src/connection.rs

```diff
@@ -210,15 +216,18 @@ impl Connection {
         mut error_sender: mpsc::UnboundedSender<MqttError>,
         done: oneshot::Sender<()>,
         permits: Arc<Semaphore>,
+        pub_count: Arc<AtomicUsize>,
     ) -> Result<(), MqttError> {
         let mut triggered_disconnect = false;
         let mut disconnect_permit = None;
+        let mut awaiting_ack = HashSet::new();
 
         loop {
             // Check if we are ready to disconnect. Due to ownership of the
             // event loop, this needs to be done before we call
             // `event_loop.poll()`
-            let remaining_events_empty = event_loop.state.inflight() == 0;
+            let remaining_events_empty =
+                event_loop.state.inflight() == 0 && pub_count.load(Ordering::SeqCst) == 0;
```

Comment:
Couldn't that be just the latter?
```suggestion
            let remaining_events_empty = pub_count.load(Ordering::SeqCst) == 0;
```

Reply:
No, because pub_count == 0 tells us that everything we intended to publish has been published, and inflight == 0 tells us that all the messages are acknowledged.